### PR TITLE
Add reset_duration option for waveshare epaper HAT rev 2.1

### LINKED ISSUE #1491: Waveshare 2.7inch e-Paper HAT shows nothing
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Hass.io, HassOS 4.13
**ESP (ESP32/ESP8266, Board/Sonoff):**

ESP32_devkitC_V4
**ESPHome version (latest production, beta, dev branch)**

1.15.1 console latest
**Affected component:**

https://esphome.io/components/display/waveshare_epaper.html
2.7inch Waveshare HAT (b/w)
https://www.waveshare.com/wiki/2.7inch_e-Paper_HAT

**Description of problem:**
- General Error : Displays shows nothing, doesnt even flicker.
- compiling without an error

**Problem-relevant YAML-configuration entries:**
```yaml
PASTE YAML FILE HERE
esphome:
  name: waveshare_display1
  platform: ESP32
  board: nodemcu-32s

wifi:
  ssid: "xxx"
  password: "xxxx"

  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Waveshare Display1"
    password: "xxx"

captive_portal:

# Enable logging
logger:
  level: debug

# Enable Home Assistant API
api:

ota:
font:
  - file: "arial.ttf"
    id: my_font
    size: 20

spi:
  clk_pin: GPIO18
  mosi_pin: GPIO23

display:
  - platform: waveshare_epaper
    cs_pin: GPIO05
    dc_pin: GPIO17
    busy_pin: GPIO04
    reset_pin: GPIO16
    model: 2.70in
    update_interval: 20s
    lambda: |-
      it.print(0, 10, id(my_font), "Hello World!");


```

**Logs (if applicable):**

```
PASTE DEBUG LOG HERE
esphome waveshare_display1.yaml run
INFO Reading configuration waveshare_display1.yaml...
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d waveshare_display1
Processing waveshare_display1 (board: nodemcu-32s; framework: arduino; platform: espressif32@1.12.4)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
PACKAGES: 
 - framework-arduinoespressif32 3.10004.200129 (1.0.4) 
 - tool-esptoolpy 1.20600.0 (2.6.0) 
 - toolchain-xtensa32 2.50200.80 (5.2.0)
Dependency Graph
|-- <AsyncTCP-esphome> 1.1.1
|-- <ESPmDNS> 1.0
|   |-- <WiFi> 1.0
|-- <FS> 1.0
|-- <ESPAsyncWebServer-esphome> 1.2.7
|   |-- <AsyncTCP-esphome> 1.1.1
|   |-- <FS> 1.0
|   |-- <WiFi> 1.0
|-- <DNSServer> 1.1.0
|   |-- <WiFi> 1.0
|-- <Update> 1.0
|-- <WiFi> 1.0
|-- <SPI> 1.0
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/user_services.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/api/util.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/display/display_buffer.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/ota/ota_component.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/spi/spi.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/waveshare_epaper/waveshare_epaper.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/web_server_base/web_server_base.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/wifi/wifi_component_esp32.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/application.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/component.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/controller.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/esphal.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/log.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/preferences.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/waveshare_display1/src/esphome/core/util.cpp.o
Compiling .pioenvs/waveshare_display1/src/main.cpp.o
Generating partitions .pioenvs/waveshare_display1/partitions.bin
Compiling .pioenvs/waveshare_display1/lib78c/AsyncTCP-esphome/AsyncTCP.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/ETH.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFi.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFiAP.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFiClient.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFiGeneric.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFiMulti.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFiSTA.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFiScan.cpp.o
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFiServer.cpp.o
Archiving .pioenvs/waveshare_display1/lib78c/libAsyncTCP-esphome.a
Indexing .pioenvs/waveshare_display1/lib78c/libAsyncTCP-esphome.a
Compiling .pioenvs/waveshare_display1/libc19/WiFi/WiFiUdp.cpp.o
Compiling .pioenvs/waveshare_display1/libd8d/ESPmDNS/ESPmDNS.cpp.o
Compiling .pioenvs/waveshare_display1/lib628/FS/FS.cpp.o
Compiling .pioenvs/waveshare_display1/lib628/FS/vfs_api.cpp.o
Compiling .pioenvs/waveshare_display1/lib3b2/ESPAsyncWebServer-esphome/AsyncEventSource.cpp.o
Compiling .pioenvs/waveshare_display1/lib3b2/ESPAsyncWebServer-esphome/AsyncWebSocket.cpp.o
Compiling .pioenvs/waveshare_display1/lib3b2/ESPAsyncWebServer-esphome/SPIFFSEditor.cpp.o
Compiling .pioenvs/waveshare_display1/lib3b2/ESPAsyncWebServer-esphome/WebAuthentication.cpp.o
Archiving .pioenvs/waveshare_display1/libc19/libWiFi.a
Indexing .pioenvs/waveshare_display1/libc19/libWiFi.a
Compiling .pioenvs/waveshare_display1/lib3b2/ESPAsyncWebServer-esphome/WebHandlers.cpp.o
Archiving .pioenvs/waveshare_display1/libd8d/libESPmDNS.a
Indexing .pioenvs/waveshare_display1/libd8d/libESPmDNS.a
Compiling .pioenvs/waveshare_display1/lib3b2/ESPAsyncWebServer-esphome/WebRequest.cpp.o
Compiling .pioenvs/waveshare_display1/lib3b2/ESPAsyncWebServer-esphome/WebResponses.cpp.o
Compiling .pioenvs/waveshare_display1/lib3b2/ESPAsyncWebServer-esphome/WebServer.cpp.o
Archiving .pioenvs/waveshare_display1/lib628/libFS.a
Indexing .pioenvs/waveshare_display1/lib628/libFS.a
Compiling .pioenvs/waveshare_display1/liba77/DNSServer/DNSServer.cpp.o
Compiling .pioenvs/waveshare_display1/lib65e/Update/Updater.cpp.o
Compiling .pioenvs/waveshare_display1/libf64/SPI/SPI.cpp.o
Archiving .pioenvs/waveshare_display1/libFrameworkArduinoVariant.a
Indexing .pioenvs/waveshare_display1/libFrameworkArduinoVariant.a
Archiving .pioenvs/waveshare_display1/libf64/libSPI.a
Indexing .pioenvs/waveshare_display1/libf64/libSPI.a
Compiling .pioenvs/waveshare_display1/FrameworkArduino/Esp.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/FunctionalInterrupt.cpp.o
Archiving .pioenvs/waveshare_display1/liba77/libDNSServer.a
Indexing .pioenvs/waveshare_display1/liba77/libDNSServer.a
Compiling .pioenvs/waveshare_display1/FrameworkArduino/HardwareSerial.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/IPAddress.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/IPv6Address.cpp.o
Archiving .pioenvs/waveshare_display1/lib65e/libUpdate.a
Indexing .pioenvs/waveshare_display1/lib65e/libUpdate.a
Compiling .pioenvs/waveshare_display1/FrameworkArduino/MD5Builder.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/Print.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/Stream.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/StreamString.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/WMath.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/WString.cpp.o
Archiving .pioenvs/waveshare_display1/lib3b2/libESPAsyncWebServer-esphome.a
Compiling .pioenvs/waveshare_display1/FrameworkArduino/base64.cpp.o
Indexing .pioenvs/waveshare_display1/lib3b2/libESPAsyncWebServer-esphome.a
Compiling .pioenvs/waveshare_display1/FrameworkArduino/cbuf.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-adc.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-bt.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-cpu.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-dac.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-gpio.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-i2c.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-ledc.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-matrix.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-misc.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-psram.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-rmt.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-sigmadelta.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-spi.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-time.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-timer.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-touch.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/esp32-hal-uart.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/libb64/cdecode.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/libb64/cencode.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/main.cpp.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/stdlib_noniso.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/wiring_pulse.c.o
Compiling .pioenvs/waveshare_display1/FrameworkArduino/wiring_shift.c.o
Archiving .pioenvs/waveshare_display1/libFrameworkArduino.a
Indexing .pioenvs/waveshare_display1/libFrameworkArduino.a
Linking .pioenvs/waveshare_display1/firmware.elf
Building .pioenvs/waveshare_display1/firmware.bin
Retrieving maximum program size .pioenvs/waveshare_display1/firmware.elf
Checking size .pioenvs/waveshare_display1/firmware.elf
RAM:   [=         ]  13.0% (used 42648 bytes from 327680 bytes)
Flash: [=====     ]  49.0% (used 898794 bytes from 1835008 bytes)
========================= [SUCCESS] Took 19.55 seconds =========================
INFO Successfully compiled program.
Found multiple options, please choose one:
  [1] /dev/ttyUSB0 (CP2102 USB to UART Bridge Controller)
  [2] Over The Air (waveshare_display1.local)
(number): 2
INFO Resolving IP address of waveshare_display1.local
INFO  -> 192.168.3.31
INFO Uploading waveshare_display1/.pioenvs/waveshare_display1/firmware.bin (898912 bytes)
Uploading: [============================================================] 100% Done...

INFO Waiting for result...
INFO OTA successful
INFO Successfully uploaded program.
INFO Starting log output from waveshare_display1.local using esphome API
INFO Connecting to waveshare_display1.local:6053 (192.168.3.31)
INFO Successfully connected to waveshare_display1.local
[18:16:41][I][app:105]: ESPHome version 1.15.1 compiled on Sep 18 2020, 18:15:56
[18:16:41][C][wifi:443]: WiFi:
[18:16:41][C][wifi:303]:   SSID: 'xxx'
[18:16:41][C][wifi:304]:   IP Address: xxxxx
[18:16:41][C][wifi:306]:   BSSID: xxxx
[18:16:41][C][wifi:307]:   Hostname: 'waveshare_display1'
[18:16:41][C][wifi:311]:   Signal strength: -59 dB ▂▄▆█
[18:16:41][V][wifi:313]:   Priority: 0.0
[18:16:41][C][wifi:315]:   Channel: 11
[18:16:41][C][wifi:316]:   Subnet: 255.255.255.0
[18:16:41][C][wifi:317]:   Gateway: xxxx
[18:16:41][C][wifi:318]:   DNS1: xxxx
[18:16:41][C][wifi:319]:   DNS2: 0.0.0.0
[18:16:41][C][spi:081]: SPI bus:
[18:16:41][C][spi:082]:   CLK Pin: GPIO18 (Mode: OUTPUT)
[18:16:41][C][spi:084]:   MOSI Pin: GPIO23 (Mode: OUTPUT)
[18:16:41][C][spi:085]:   Using HW SPI: YES
[18:16:41][C][logger:185]: Logger:
[18:16:41][C][logger:186]:   Level: VERBOSE
[18:16:41][C][logger:187]:   Log Baud Rate: 115200
[18:16:41][C][logger:188]:   Hardware UART: UART0
[18:16:41][C][waveshare_epaper:451]: Waveshare E-Paper
[18:16:41][C][waveshare_epaper:451]:   Rotations: 0 °
[18:16:41][C][waveshare_epaper:451]:   Dimensions: 176px x 264px
[18:16:41][C][waveshare_epaper:452]:   Model: 2.7in
[18:16:41][C][waveshare_epaper:453]:   Reset Pin: GPIO16 (Mode: OUTPUT)
[18:16:41][C][waveshare_epaper:454]:   DC Pin: GPIO17 (Mode: OUTPUT)
[18:16:41][C][waveshare_epaper:455]:   Busy Pin: GPIO4 (Mode: INPUT)
[18:16:41][C][waveshare_epaper:456]:   Update Interval: 20.0s
[18:16:41][C][captive_portal:169]: Captive Portal:
[18:16:41][C][ota:029]: Over-The-Air Updates:
[18:16:41][C][ota:030]:   Address: waveshare_display1.local:3232
[18:16:41][C][api:095]: API Server:
[18:16:41][C][api:096]:   Address: waveshare_display1.local:6053
```

**Additional information and things you've tried:**

- [ ]  compiled examplie sketch with Arduino IDE with the same connection - workes

- [ ]  compiled expample with python with a raspberry pi according to the description of the manufactor - runs like a charm
- [ ] mesured the voltage of the display and the board
- [ ]  Tried different GPIOS according to https://components101.com/sites/default/files/component_pin/ESP32-Pinout.png
- [ ] different Voltage (5V)
- [ ] tried same setup with ESP12e - no success
- [ ] tried simple line command `it.line(0, 0, 50, 50);` for lambda
- [ ] checked cable connection with multimeter

- [ ] additional serial log : [serial_veryverboselog Pastebin](https://pastebin.pl/view/e93b7bd0)

### PR DESCRIPTION
**Pull request by owenb321**
## Description:

Add reset_duration option to Waveshare display to support newer e-Paper driver modules (Rev 2.1)

**Related issue (if applicable):** May fix esphome/issues#1491

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#952

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by damanti-me** [comment]
thank you! This resolved my issue esphome/issues#1491

**Comment by damanti-me** [comment]
Hey, will you resolve the conflicting files soon? Id like to have this reset_duration functionality in the dev version... Thanks

**Comment by frenck** [comment]
Hi there @owenb321!

Friendly reminder on this PR! There is a merge conflict, could you take a look and resolve that? 

Thanks already! 👍 

../Frenck 🚂

**Comment by hugokernel** [comment]
Hi,
This PR solves my problem on a Waveshare 7in5v2 (screen that displays nothing) with a reset duration of 2 (like original Arduino for V2 and HD versions).
Thank you @owenb321

**Comment by hugokernel** [comment]
Any news about this PR ? Thx

**Comment by beniwohli** [comment]
This also made a 7.50inv2 display connected through a e-Paper Driver HAT **rev 2.2** work (also using a `reset_duration` of `2`).

**Comment by brase** [comment]
How do I test this? I have already built esphome from this branch but it says:

```
  [reset_duration] is an invalid option for [display.waveshare_epaper]. Did you mean [reset_pin], [rotation]?
```

Has anyone an idea?

**Comment by Foamier** [comment]
Hi all,
Struggled for a long time to get the Waveshare 7.50inV2 black e-paper working in ESPhome. Arduino is easy compared to this, going with the Waveshare examples.

until the merge of this option is completed and added into the main release, I've done a quick change on ESPhome to lower the reset delay to 2ms (is this milli-seconds?).
Howto:
- install Portainer as an addon
- in Portainer get onto the console of addon_xxxxxxxx_ESPhome container.
- Change the source file in ESPhome. This is needed since every compilation takes a new copy of this file.
sed -i 's/delay(200);/delay(2);/' /opt/esphome/esphome/components/waveshare_epaper/waveshare_epaper.h
- Keep in mind this is for all Waveshare e-paper displays when you compile in ESPhome. I have not tested this with other epapers.
- As soon as ESPhome is updated or restarted, the changes are gone and need to be re-applied.

If someone has a better way to link a different waveshare_epaper.h to an ESPhome yaml file (a method that does not require a great deal of skill), that would be better of course and would love to hear about it.

File: tests/test1.yaml [hunk 1]
```diff
@@ -1922,6 +1922,16 @@ display:
...
+  - platform: waveshare_epaper
+    cs_pin: GPIO23
+    dc_pin: GPIO23
+    busy_pin: GPIO23
+    reset_pin: GPIO23
+    model: 2.90in
+    full_update_every: 30
+    reset_duration: 200
+    lambda: |-
+      it.rectangle(0, 0, it.get_width(), it.get_height());
```
Comment (jesserockz) [inline tests/test1.yaml]: Please move this to test4 where the other waveshare tests are.

File: esphome/components/waveshare_epaper/display.py [hunk 2]
```diff
@@ -81,6 +82,7 @@ def validate_full_update_every_only_type_a(value):
             cv.Optional(CONF_RESET_PIN): pins.gpio_output_pin_schema,
             cv.Optional(CONF_BUSY_PIN): pins.gpio_input_pin_schema,
             cv.Optional(CONF_FULL_UPDATE_EVERY): cv.uint32_t,
+            cv.Optional(CONF_RESET_DURATION): cv.uint32_t,
```
Comment (jesserockz) [inline esphome/components/waveshare_epaper/display.py]: This should be a `cv.positive_time_period_milliseconds`
```suggestion
            cv.Optional(CONF_RESET_DURATION): cv.positive_time_period_milliseconds,
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**